### PR TITLE
Flume input operator and agent

### PR DESCRIPTION
**Pull request by pramodin**
This PR adds support for flume. It contains flume input operator and an agent to use on the flume side.

### CODE REVIEW

**Comment by pramodin** [comment]
This is in a cleaned up format and contains documentation and tests as discussed in the earlier PR https://github.com/apache/apex-malhar/pull/475

**Comment by pramodin** [comment]
@tweise, others please see

**Comment by pramodin** [comment]
Rebase to latest. @tweise, others who are interested please review

**Comment by tweise** [comment]
@PramodSSImmaneni there are a few more replacements that should be done, for example

DT -> Apex 
.dt -> .apex

File: flume/pom.xml [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+                    <mapping>
+                      <directory>${package.prefix}/flume-${project.version}/lib</directory>
+                      <filemode>750</filemode>
+                      <username>${package.username}</username>
+                      <groupname>${package.groupname}</groupname>
+                      <artifact></artifact>
+                      <dependency>
+                        <includes>
+                          <include>org.apache.apex:apex-api:jar:${apex.core.version}</include>
+                          <include>com.datatorrent:dt-netlet:jar:1.2.0</include>
```
Comment (tweise) [inline flume/pom.xml]: Is it necessary to repeat/specify the version number here? This will break when we update core.

**Comment by pramodin** [comment]
How about letting the remaining DT references be. Don't want to have to go through one more iteration of testing. They seem to mainly in sample files (which the user can replace with their own names) with an exception of one of two classes having DT in the name. Will look at the netlet reference in the rpm packaging.

**Comment by pramodin** [comment]
@tweise updated dependency and rebased

**Comment by tweise** [comment]
The DT references are 2 class renames that your IDE does for you and a couple of replacements with sed, that don't require any testing. No need to change sample files.

```
$ git grep DT
README.md:DT_FLUME_JAR must point to the plugin JAR.)
README.md:In the configuration file set `org.apache.apex.malhar.flume.sink.DTFlumeSink` for the **type**  
src/main/java/org/apache/apex/malhar/flume/discovery/Discovery.java: * When DTFlumeSink server instance binds to the network interface, it
src/main/java/org/apache/apex/malhar/flume/discovery/Discovery.java: * they can find an available DTFlumeSink server instance.
src/main/java/org/apache/apex/malhar/flume/discovery/Discovery.java:   * Advertise the host/port address where DTFlumeSink is accepting a 
src/main/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscovery.java:    this.serviceName = "DTFlume";
src/main/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscovery.java:    serviceName = context.getString("serviceName", "DTFlume"
src/main/java/org/apache/apex/malhar/flume/operator/AbstractFlumeInputOperator.java:      logger.info("Discovered zero DTFlumeSink");
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java: * DTFlumeSink is a flume sink developed to ingest the data into DataTorr
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java: * &lt;experimental&gt;DTFlumeSink auto adjusts the rate at which it cons
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java: * The properties you can set on the DTFlumeSink are: <br />
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java:public class DTFlumeSink extends AbstractSink implements Configurable
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java:      logger.warn("storage key missing... DTFlumeSink may lose data!");
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java:  private static final Logger logger = LoggerFactory.getLogger(DTFlumeSin
src/main/resources/flume-conf/flume-conf.sample.properties: agent1.sinks.dt.type = org.apache.apex.malhar.flume.sink.DTFlumeSink
src/main/resources/flume-conf/flume-env.sample.sh:if test -z "$DT_FLUME_JAR"
src/main/resources/flume-conf/flume-env.sample.sh:  echo [ERROR]: Environment variable DT_FLUME_JAR should point to a valid jar file which
README.md:DT_FLUME_JAR must point to the plugin JAR.)
README.md:In the configuration file set `org.apache.apex.malhar.flume.sink.DTFlumeSink` for the **type**  
src/main/java/org/apache/apex/malhar/flume/discovery/Discovery.java: * When DTFlumeSink server instance binds to the network interface, it
src/main/java/org/apache/apex/malhar/flume/discovery/Discovery.java: * they can find an available DTFlumeSink server instance.
src/main/java/org/apache/apex/malhar/flume/discovery/Discovery.java:   * Advertise the host/port address where DTFlumeSink is accepting a 
src/main/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscovery.java:    this.serviceName = "DTFlume";
src/main/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscovery.java:    serviceName = context.getString("serviceName", "DTFlume"
src/main/java/org/apache/apex/malhar/flume/operator/AbstractFlumeInputOperator.java:      logger.info("Discovered zero DTFlumeSink");
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java: * DTFlumeSink is a flume sink developed to ingest the data into DataTorr
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java: * &lt;experimental&gt;DTFlumeSink auto adjusts the rate at which it cons
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java: * The properties you can set on the DTFlumeSink are: <br />
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java:public class DTFlumeSink extends AbstractSink implements Configurable
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java:      logger.warn("storage key missing... DTFlumeSink may lose data!");
src/main/java/org/apache/apex/malhar/flume/sink/DTFlumeSink.java:  private static final Logger logger = LoggerFactory.getLogger(DTFlumeSin
src/main/resources/flume-conf/flume-conf.sample.properties: agent1.sinks.dt.type = org.apache.apex.malhar.flume.sink.DTFlumeSink
src/main/resources/flume-conf/flume-env.sample.sh:if test -z "$DT_FLUME_JAR"
src/main/resources/flume-conf/flume-env.sample.sh:  echo [ERROR]: Environment variable DT_FLUME_JAR should point to a valid jar file which
src/main/resources/flume-conf/flume-env.sample.sh:FLUME_CLASSPATH=`JARPATH=$JARPATH $JAVA -cp $DT_FLUME_JAR com.datatorrent.jarpath.JarPat
src/test/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscoveryTest.java:    discovery.setServiceName("DTFlumeTest");
src/test/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscoveryTest.java:    discovery.setBasePath("/HelloDT");
src/test/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscoveryTest.java:    discovery.setServiceName("DTFlumeTest");
src/test/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscoveryTest.java:    discovery.setBasePath("/HelloDT");
src/test/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscoveryTest.java:    discovery.setServiceName("DTFlumeTest");
src/test/java/org/apache/apex/malhar/flume/discovery/ZKAssistedDiscoveryTest.java:    discovery.setBasePath("/HelloDT");
src/test/java/org/apache/apex/malhar/flume/sink/DTFlumeSinkTest.java:public class DTFlumeSinkTest
src/test/java/org/apache/apex/malhar/flume/sink/DTFlumeSinkTest.java:    DTFlumeSink sink = new DTFlumeSink();
src/test/java/org/apache/apex/malhar/flume/sink/DTFlumeSinkTest.java:        synchronized (DTFlumeSinkTest.this) {
src/test/java/org/apache/apex/malhar/flume/sink/DTFlumeSinkTest.java:          DTFlumeSinkTest.this.notify();
src/test/java/org/apache/apex/malhar/flume/sink/DTFlumeSinkTest.java:  private static final Logger logger = LoggerFactory.getLogger(DTFlum
src/test/resources/flume/conf/flume-conf.properties:agent1.sinks.dt.type = org.apache.apex.malhar.flume.sink.DTFlumeSink
src/test/resources/flume/conf/flume-conf.properties:   agent1.sinks.dt.discovery.basePath = /HelloDT
src/test/resources/flume/conf/flume-env.sh:if test -z "$DT_FLUME_JAR"
src/test/resources/flume/conf/flume-env.sh:  echo [ERROR]: Environment variable DT_FLUME_JAR should point to a valid jar file which contai
src/test/resources/flume/conf/flume-env.sh:FLUME_CLASSPATH=`JARPATH=$JARPATH $JAVA -cp $DT_FLUME_JAR com.datatorrent.jarpath.JarPath -N $D
src/test/resources/flume/conf/flume_simple.conf: a1.sinks.dt.type = org.apache.apex.malhar.flume.sink.DTFlumeSink
src/test/resources/flume/conf/flume_zkdiscovery.conf: a1.sinks.dt.type = org.apache.apex.malhar.flume.sink.DTFlumeSink
src/test/resources/flume/conf/flume_zkdiscovery.conf: a1.sinks.dt2.type = org.apache.apex.malhar.flume.sink.DTFlumeSink
```

**Comment by pramodin** [comment]
Done.

**Comment by tweise** [comment]
LGTM, please rebase and see that CI passes. There seem to be intermittent errors not caused by this PR.

**Pull request merged**